### DISCUSSION TITLE
Are input/output clusters ordered?

### DISCUSSION THREAD

**Discussion by mkolesnik** [zigpy/zigpy#1150] [Q&A]
When declaring `INPUT_CLUSTERS` and `OUTPUT_CLUSTERS` they're declared as lists, but in many places are treated as sets.

Does the ordering in the declaration matter, or is it just the contents that matter?

Specifically I have a curtain module `TS130F by _TZ3000_zirycpws` which gets "misidentified" as a `TuyaTS130GP` and misbehaving (might not be related to this though), and I'm wondering if the ordering of the endpoint clusters matters or not (I think not, but want to be sure).

**Comment by puddly** [comment]
The order is ignored:

https://github.com/zigpy/zigpy/blob/297a2b0f364fe7a31de3b4966d70d1a67d432358/zigpy/quirks/registry.py#L153-L155

**Reply by mkolesnik** [reply]
For the sake of matching yes, but for the sake of the actual operations is it also ignored?

I guess for the replacements both lists have to match, but for sending/receiving the actual command does it matter?

**Reply by puddly** [reply]
Not at all. If you have a device whose signature matches another one's *exactly*, that's a little bit of an issue, as it seems like Tuya reused a randomly-generated device ID...
